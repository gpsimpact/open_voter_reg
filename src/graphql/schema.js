import glob from 'glob';
import fs from 'fs';
import path from 'path';
import { makeExecutableSchema } from 'graphql-tools';

process.on('warning', e => console.warn(e.stack));

const typeDefs = [];

// loop over types and import the contents
glob.sync('**/*.type.gql', { cwd: __dirname }).forEach(filename => {
  const filePath = path.join(__dirname, filename);
  const fileContent = fs.readFileSync(filePath, { encoding: 'utf8' });
  typeDefs.push(fileContent);
});

import resolvers from './resolvers';

// create a schema
const schema = makeExecutableSchema({ typeDefs, resolvers });

// add mocks to the schema, preserving the existing resolvers (none for the time beeing)
// import mocks from './mocks';
// addMockFunctionsToSchema({ schema, mocks, preserveResolvers: true });

export default schema;
