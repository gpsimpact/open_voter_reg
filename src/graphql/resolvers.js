import QueryResolver from './Query/Query.resolvers';
import MutationResolver from './Mutation/Mutation.resolvers';

export default {
  ...QueryResolver,
  ...MutationResolver,
};
