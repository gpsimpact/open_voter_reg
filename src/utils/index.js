import faker from 'faker';
import { includes } from 'lodash';
import { InsufficientPermissionsError } from '../errors';

export function mapTo(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), row));
    return Array.from(group.values());
  };
}

export function mapToMany(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, []]));
    rows.forEach(row => (group.get(keyFn(row)) || []).push(row));
    return Array.from(group.values());
  };
}

export function mapToValues(keys, keyFn, valueFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), valueFn(row)));
    return Array.from(group.values());
  };
}

export const generateFakeUsers = (num, seed) => {
  faker.seed(seed); // ensures consistent result
  const fakeUsers = [];
  Array(num)
    .fill()
    .map((_, i) => {
      fakeUsers.push({
        first_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
      });
    });
  return fakeUsers;
};

export const hasPermission = (viewer, service, requiredPermission, throwOnFail = false) => {
  if (!includes(viewer.permissions[service], requiredPermission)) {
    if (throwOnFail) {
      throw new InsufficientPermissionsError();
    }
    return false;
  }
  return true;
};
