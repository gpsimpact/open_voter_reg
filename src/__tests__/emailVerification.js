import { graphql } from 'graphql';
import db from '../db';
import schema from '../graphql/schema';
// import faker from 'faker';
import MakeContext from '../Context';
import { find } from 'lodash';
// import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { generateFakeUsers } from '../utils';

beforeAll(async () => await db.migrate.latest({ directory: './src/db/migrations' }));
beforeEach(async () => await Promise.all([db.raw('TRUNCATE TABLE users CASCADE')]));
afterAll(async () => await db.destroy());

describe('email verification', () => {
  test('mutation should trigger an email with a predictable link and JWT token in querystring', async () => {
    const sendEmail = jest.fn(() => Promise.resolve());
    const users = generateFakeUsers(1, 1);
    const emailVerificationToken = jwt.sign({ email: users[0].email }, process.env.JWT_SECRET);
    const base_url = 'https://facebook.com';
    const query = `
      mutation {
        sendVerificationEmail(email:"${users[0].email}", base_url: "${base_url}")
      }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: null });
    context.connectors.sendEmail = sendEmail;
    await graphql(schema, query, rootValue, context);
    expect(sendEmail).toHaveBeenCalled();
    expect(sendEmail.mock.calls[0][0].to).toBe(users[0].email);
    expect(sendEmail.mock.calls[0][0].html).toEqual(
      expect.stringContaining(`${base_url}/emailVerification?token=${emailVerificationToken}`)
    );
    expect(sendEmail.mock.calls[0][0].txt).toEqual(
      expect.stringContaining(`${base_url}/emailVerification?token=${emailVerificationToken}`)
    );
  });

  test('mutation should mark email as verified', async () => {
    const users = generateFakeUsers(1, 1);
    await db('users').insert({
      email: users[0].email,
      email_verified: false,
    });
    const emailVerificationToken = jwt.sign({ email: users[0].email }, process.env.JWT_SECRET);
    const query = `
      mutation {
        verifyEmailAddress(token:"${emailVerificationToken}")
      }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    expect(result.data.verifyEmailAddress).toBe('ok');
    const dbUserRecord = await db('users')
      .first()
      .where({ email: users[0].email });
    expect(dbUserRecord.email_verified).toBe(true);
  });

  test('it should fail if token is bad', async () => {
    const query = `
      mutation {
        verifyEmailAddress(token:"badToken.blergh")
      }
    `;
    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    // console.log(JSON.stringify(result, null, '\t'));
    expect(result).toMatchSnapshot();
    expect(result.errors.length).toBeGreaterThanOrEqual(1);
    expect(find(result.errors, { message: 'Invalid Token.' })).not.toBeUndefined();
  });
});
