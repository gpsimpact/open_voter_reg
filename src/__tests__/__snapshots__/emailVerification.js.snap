// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email verification it should fail if token is bad 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Invalid Token.],
  ],
}
`;
